src/sdam/topology.rs

```diff
@@ -415,173 +393,139 @@ impl TopologyWorker {
         });
     }
 
-    /// Get a reference to the latest topology state.
-    ///
-    /// Note that this holds a read lock, so broadcasting a new state while an
-    /// outstanding reference to the `TopologyState` is held will deadlock.
-    fn borrow_latest_state(&self) -> Ref<TopologyState> {
-        self.topology_watcher.peek_latest()
+    /// Publish the current TopologyDescription and map of Servers.
+    fn publish_state(&self) {
+        let servers = self
+            .servers
+            .iter()
+            .map(|(k, v)| (k.clone(), Arc::downgrade(v)))
+            .collect();
+        self.publisher.publish_new_state(TopologyState {
+            description: self.topology_description.clone(),
+            servers,
+        })
     }
 
     fn advance_cluster_time(&mut self, to: ClusterTime) {
-        let mut latest_state = self.borrow_latest_state().clone();
-        latest_state.description.advance_cluster_time(&to);
-        self.publisher.publish_new_state(latest_state);
+        self.topology_description.advance_cluster_time(&to);
+        self.publish_state()
     }
 
-    fn sync_hosts(&self, state: &mut TopologyState, hosts: HashSet<ServerAddress>) {
-        state.servers.retain(|host, _| hosts.contains(host));
-        state.description.sync_hosts(&hosts);
-
-        for address in hosts {
-            if state.servers.contains_key(&address) {
-                continue;
-            }
-            let server = Server::new(
-                address.clone(),
-                self.options.clone(),
-                self.http_client.clone(),
-                self.topology_updater.clone(),
-            );
-            state.servers.insert(address.clone(), server);
-
-            #[cfg(test)]
-            let disable_monitoring_threads = self
-                .options
-                .test_options
-                .as_ref()
-                .map(|to| to.disable_monitoring_threads)
-                .unwrap_or(false);
-            #[cfg(not(test))]
-            let disable_monitoring_threads = false;
-
-            if !disable_monitoring_threads {
-                Monitor::start(
-                    address,
-                    self.topology_updater.clone(),
-                    self.topology_watcher.clone(),
-                    self.event_emitter.clone(),
-                    self.update_requester.subscribe(),
-                    self.options.clone(),
-                );
-            }
-        }
+    fn sync_hosts(&mut self, hosts: HashSet<ServerAddress>) -> bool {
+        let mut new_description = self.topology_description.clone();
+        new_description.sync_hosts(&hosts);
+        self.update_topology(new_description)
```

Comment:
Everything that modifies the topology now goes through the following steps:
1. clone the existing topology description
2. modify it in some way
3. pass it to `update_topology`

This cuts down on the one-off logic in these various methods and hopefully makes the worker easier to read / reason about.